hivemq-edge/src/frontend/src/modules/Workspace/components/nodes/ContextualToolbar.tsx

```diff
@@ -0,0 +1,139 @@
+import { type FC, MouseEventHandler, useMemo } from 'react'
+import { NodeToolbar, type NodeProps, type NodeToolbarProps, Position, Node, Edge, MarkerType } from 'reactflow'
+import { useTranslation } from 'react-i18next'
+import { useTheme } from '@chakra-ui/react'
+import { LuPanelRightOpen } from 'react-icons/lu'
+import { ImMakeGroup } from 'react-icons/im'
+
+import { Adapter, Status } from '@/api/__generated__'
+import { EdgeTypes, Group, IdStubs, NodeTypes } from '@/modules/Workspace/types.ts'
+import IconButton from '@/components/Chakra/IconButton.tsx'
+import useWorkspaceStore from '@/modules/Workspace/hooks/useWorkspaceStore.ts'
+import { getGroupLayout } from '@/modules/Workspace/utils/group.utils.ts'
+import { getThemeForStatus } from '@/modules/Workspace/utils/status-utils.ts'
+import WorkspaceButtonGroup from '@/modules/Workspace/components/parts/WorkspaceButtonGroup.tsx'
+
+type SelectedNodeProps = Pick<NodeProps, 'id'> & Pick<NodeToolbarProps, 'position'>
+interface ContextualToolbarProps extends SelectedNodeProps {
+  onOpenPanel?: MouseEventHandler | undefined
+  children?: React.ReactNode
+}
+
+const ContextualToolbar: FC<ContextualToolbarProps> = ({ id, onOpenPanel, children }) => {
+  const { t } = useTranslation()
+  const { onInsertGroupNode, nodes } = useWorkspaceStore()
+  const theme = useTheme()
+
+  const selectedNodes = nodes.filter((node) => node.selected)
+  const selectedGroupCandidates = useMemo(() => {
+    // TODO[NVL] Should the grouping only be available if ALL nodes match the filter ?
```

Comment:
That would be an interesting one to user test. My guess would be that a user would expect grouping to be available if at least 2 nodes are selected, as per your code. Hence, if I've experienced grouping before, I could get frustrated if grouping only was available if all nodes were selected. And how would I know why grouping isn't available?

Reply:
The grouping affordance is skewed when selectable nodes are not suitable (e.g. already in another group or of a different type). 
I'm relying on a `select-then-action` paradigm, using the "global" selection mechanism. Hence some nodes might NOT end up in your group. 
A better alternative would be to implement `action-then-select` paradigm: since you are in a "grouping" context, only nodes that are suitable for grouping can be selected. It's a bit more difficult to implement.